### ISSUE TITLE
[BUG] SMP signal dispatch seems racy

### ISSUE THREAD

**Issue by yamt** [apache/nuttx#14807]
### Description / Steps to reproduce the issue

the signal sending logic (kill) basically does:
1. enter_critical_section
2. decide the target thread
3. check signal mask of the thread
4. make a cross call to the target cpu (nxsched_smp_call_single)

as nxsched_smp_call_single drops the critical section, there seems to be nothing to prevent the signal mask from changing in the meantime. it can end up with a signal delivered while being blocked.

similarly, the need_restore logic in sig_handler seems dangerous as it can revert unrelated changes to those fields.

### On which OS does this issue occur?

[OS: Mac]

### What is the version of your OS?

macOS 14.7

### NuttX Version

master

### Issue Architecture

[Arch: all]

### Issue Area

[Area: Kernel]

### Verification

- [X] I have verified before submitting the report.

**Comment by xiaoxiang781216** [comment]
@hujun260 please look at this issue.

**Comment by hujun260** [comment]
> as nxsched_smp_call_single drops the critical section, there seems to be nothing to prevent the signal mask from changing in the meantime. it can end up with a signal delivered while being blocked.

I don't think this change caused https://github.com/apache/nuttx/issues/14749, 
because the release of the critical section is going to happen regardless of whether I call nxsched_smp_call_single or not.
![image](https://github.com/user-attachments/assets/103070aa-35b1-4835-8285-e163ae067945)


> similarly, the need_restore logic in sig_handler seems dangerous as it can revert unrelated changes to those fields.

There are two fields: affinity and flag. 
Due to the setting of TCB_FLAG_CPU_LOCKED, the affinity will not be changed during the process. 
The tcb->flags may undergo changes, which I can further refine. During restoration, only the TCB_FLAG_CPU_LOCKED field of tcb->flags will be restored, without altering any other fields.

**Comment by yamt** [comment]
> > as nxsched_smp_call_single drops the critical section, there seems to be nothing to prevent the signal mask from changing in the meantime. it can end up with a signal delivered while being blocked.
> 
> I don't think this change caused #14749, because the release of the critical section is going to happen regardless of whether I call nxsched_smp_call_single or not. ![image](https://private-user-images.githubusercontent.com/128452594/386530328-103070aa-35b1-4835-8285-e163ae067945.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE2NzU3MjksIm5iZiI6MTczMTY3NTQyOSwicGF0aCI6Ii8xMjg0NTI1OTQvMzg2NTMwMzI4LTEwMzA3MGFhLTM1YjEtNDgzNS04Mjg1LWUxNjNhZTA2Nzk0NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNVQxMjU3MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iODg1Mjk2MTU4NWY2NDM4NTU1Nzg0ZmYyNTgzZjRkYTI5ZGIzM2QyMzFlMDZiMmFhYTE5ZjZjM2ZjOGQ5YmY4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.J_hhPdrBL-n4uzjaJRCAgq3-6JJQmGI9HEeP3qDq4k4)

i don't understand your logic.
in non cross call case, the critical section is released after up_schedule_sigaction.

**Comment by hujun260** [comment]
> in non cross call case, the critical section is released after up_schedule_sigaction.

Regardless of whether it crosses cores, up_schedule_sigaction will be invoked, and nxsched_smp_call_single will wait until up_schedule_sigaction has been called and returns.

**Comment by yamt** [comment]
> > in non cross call case, the critical section is released after up_schedule_sigaction.
> 
> Regardless of whether it crosses cores, up_schedule_sigaction will be invoked, and nxsched_smp_call_single will wait until up_schedule_sigaction has been called and returns.

a cross call releases the critical section before calling up_schedule_sigaction. thus racy.

**Comment by hujun260** [comment]
> a cross call releases the critical section before calling up_schedule_sigaction. thus racy.

The vulnerability appears to be persistent, even if up_schedule_sigaction is called immediately in critical section,
it still releases the critical section immediately and cannot prevent the signal mask from changing

**Comment by yamt** [comment]
> > a cross call releases the critical section before calling up_schedule_sigaction. thus racy.
> 
> The vulnerability appears to be persistent, even if up_schedule_sigaction is called immediately in critical section, it still releases the critical section immediately and cannot prevent the signal mask from changing

if you mean "it isn't trivial to fix", i agree.

**Comment by hujun260** [comment]
> > > a cross call releases the critical section before calling up_schedule_sigaction. thus racy.
> > 
> > 
> > The vulnerability appears to be persistent, even if up_schedule_sigaction is called immediately in critical section, it still releases the critical section immediately and cannot prevent the signal mask from changing
> 
> if you mean "it isn't trivial to fix", i agree.

yes
